### PR TITLE
Configure cache sizes based on cluster size

### PR DESCRIPTION
**Pull request by wojtek-t**
@fabioy

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29983)

### CODE REVIEW

**Comment by wojtek-t** [comment]
@lavalamp - friendly ping

File: pkg/registry/configmap/etcd/etcd.go [hunk 1]
```diff
@@ -36,7 +37,13 @@ func NewREST(opts generic.RESTOptions) *REST {
 
 	newListFunc := func() runtime.Object { return &api.ConfigMapList{} }
 	storageInterface := opts.Decorator(
-		opts.Storage, 100, &api.ConfigMap{}, prefix, configmap.Strategy, newListFunc, storage.NoTriggerPublisher)
+		opts.Storage,
+		cachesize.GetWatchCacheSizeByResource(cachesize.ConfigMaps),
```
Comment (deads2k) [inline pkg/registry/configmap/etcd/etcd.go]: Seems like you really want this to be an option on the `RESTOptions`.  Those are already built per-resource.

File: pkg/registry/cachesize/cachesize.go [hunk 2]
```diff
@@ -56,40 +58,27 @@ const (
...
-	watchCacheSizes[ResourceQuotas] = 100
-	watchCacheSizes[ScheduledJobs] = 100
-	watchCacheSizes[Roles] = 100
-	watchCacheSizes[RoleBindings] = 100
-	watchCacheSizes[Secrets] = 100
-	watchCacheSizes[ServiceAccounts] = 100
-	watchCacheSizes[Services] = 100
+}
+
+func InitializeWatchCacheSizes(clusterSize int) {
```
Comment (fabioy) [inline pkg/registry/cachesize/cachesize.go]: A typical pattern for large clusters is to create a small one, test it, then resize it up. Unless there's a dynamic way to adjust these values, initializing it once doesn't seem to be a great strategy.

A suggestion: given that GKE will resize the master based upon the current cluster size, why not use the total physical memory on the machine to set these parameters?

Comment (wojtek-t) [inline pkg/registry/cachesize/cachesize.go]: @fabioy - when I asked this question to @roberthbailey he said that when you are resizing masters, you will recompute all the parameters that are passed via salt/gci/etc.; it that's the case, it would be enough for us.
Or is it not the case?

File: pkg/registry/configmap/etcd/etcd.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/registry/configmap/etcd/etcd.go]: I just unified it for the remaining 3 resources in this PR (we already have this pattern for other 20 or so).
I'm happy to change that, but I would prefer to do it in a separate PR.

Comment (deads2k) [inline pkg/registry/configmap/etcd/etcd.go]: > I just unified it for the remaining 3 resources in this PR (we already have this pattern for other 20 or so).
> I'm happy to change that, but I would prefer to do it in a separate PR

Ok

File: pkg/genericapiserver/options/server_run_options.go [hunk 3]
```diff
@@ -71,6 +71,7 @@ type ServerRunOptions struct {
 	EnableSwaggerUI           bool
 	EnableWatchCache          bool
 	EtcdServersOverrides      []string
+	ExpectedClusterSize       int
```
Comment (lavalamp) [inline pkg/genericapiserver/options/server_run_options.go]: This is really really hard to take. apiserver should be *publishing* this value, not accepting it as a command line argument.

Comment (lavalamp) [inline pkg/genericapiserver/options/server_run_options.go]: A flag I could live with might be TargetRAMUsage or some such.

File: pkg/registry/cachesize/cachesize.go (hunk 2 above)
Comment (fabioy) [inline pkg/registry/cachesize/cachesize.go]: Oh, I see, you were going to rely on that. That'd work too, as long as the bucket boundaries are ok with you.

Comment (wojtek-t) [inline pkg/registry/cachesize/cachesize.go]: Yes - bucketing is good enough for that purpose.

File: pkg/genericapiserver/options/server_run_options.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/genericapiserver/options/server_run_options.go]: Got your point - will fix tomorrow. Thanks!

File: pkg/registry/cachesize/cachesize.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/registry/cachesize/cachesize.go]: It seems that once I apply Daniel`s comment, it will be more bucketed anyway. So that would work perfectly with that.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -619,6 +619,10 @@ function start-kube-apiserver {
   if [[ -n "${ENABLE_GARBAGE_COLLECTOR:-}" ]]; then
     params+=" --enable-garbage-collector=${ENABLE_GARBAGE_COLLECTOR}"
   fi
+  if [[ -n "${NUM_NODES:-}" ]]; then
+    # Heuristic used to compute machine capacity based on number of nodes.
+    params+=" --expected-ram-capacity-mb=$((${NUM_NODES} * 60))"
```
Comment (wojtek-t) [inline cluster/gce/gci/configure-helper.sh]: @lavalamp - this is some heuristic;
Soon we should start setting up proper requests and limits for our components and we will reuse those here.

File: pkg/genericapiserver/options/server_run_options.go [hunk 5]
```diff
@@ -73,6 +73,7 @@ type ServerRunOptions struct {
 	EnableSwaggerUI           bool
 	EnableWatchCache          bool
 	EtcdServersOverrides      []string
+	ExpectedRAMCapacityMB     int
```
Comment (lavalamp) [inline pkg/genericapiserver/options/server_run_options.go]: Hrm, this is also not great. Maybe "TargetRAMUsageMB"?

"Expected" is a weird word here, the system should definitely know how much ram there is before it runs this.

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (lavalamp) [inline cluster/gce/gci/configure-helper.sh]: OK, please make that clearer in the comment (all places the comment appears)?

File: pkg/genericapiserver/options/server_run_options.go [hunk 6]
```diff
@@ -304,6 +305,9 @@ func (s *ServerRunOptions) AddFlags(fs *pflag.FlagSet) {
 		"Per-resource etcd servers overrides, comma separated. The individual override "+
 		"format: group/resource#servers, where servers are http://ip:port, semicolon separated.")
 
+	fs.IntVar(&s.ExpectedRAMCapacityMB, "expected-ram-capacity-mb", s.ExpectedRAMCapacityMB,
+		"Expected RAM capacity for apiserver in MB (used to configure sizes of caches, etc.)")
```
Comment (lavalamp) [inline pkg/genericapiserver/options/server_run_options.go]: s/capacity/available/

..."limit" or "request" would be good words to work in, since they're how we talk about resource usage in config files...

**Comment by wojtek-t** [comment]
Comments applied.

Self-applying lgtm label based on offline conversation with @lavalamp - thanks for the review!
